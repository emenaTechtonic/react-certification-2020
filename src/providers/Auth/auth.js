import React, { useState, useEffect, createContext, useCallback } from 'react'
import { AUTH_STORAGE_KEY, AUTH_STORAGE_USER } from '../../utils/constants'
import { storage } from '../../utils/storage'

export const AuthContext = createContext(null)

export const AuthProvider = ({ children }) => {
  const [authenticated, setAuthenticated] = useState(false)

  useEffect(() => {
    const lastAuthState = storage.get(AUTH_STORAGE_KEY)
    const isAuthenticated = Boolean(lastAuthState)

    setAuthenticated(isAuthenticated)
  }, [])

  const login = useCallback((user) => {
    setAuthenticated(true)
    storage.set(AUTH_STORAGE_KEY, true)
    storage.set(AUTH_STORAGE_USER, user)
  }, [])

  const logout = useCallback((user) => {
    setAuthenticated(false)
    storage.set(AUTH_STORAGE_KEY, false)
    storage.set(AUTH_STORAGE_USER, user)
  }, [])

  return (
    <AuthContext.Provider value={{ login, logout, authenticated }}>
      {children}
    </AuthContext.Provider>
  )
}

// export { useAuth }
export default AuthProvider
