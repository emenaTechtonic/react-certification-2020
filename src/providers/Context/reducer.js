export const ContextReducer = (state, action) => {
  switch (action.type) {
    case 'TOGGLE_THEME':
      return {
        ...state,
        theme: !state.theme,
      }
    case 'OPEN_MENU':
      return {
        ...state,
        menu: action.payload,
      }
    case 'GET_DATA':
      return {
        ...state,
        data: action.payload,
      }
    case 'SET_VIEW':
      return {
        ...state,
        view: action.payload,
      }
    case 'SET_TARGET':
      return {
        ...state,
        target: action.payload,
      }
    case 'SET_USER':
      return {
        ...state,
        user: action.payload,
      }
    case 'SET_FAVORITES':
      return {
        ...state,
        favorites: action.payload,
      }

    default:
      return state
  }
}
