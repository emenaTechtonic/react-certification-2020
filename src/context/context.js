import React, { useEffect, createContext, useReducer, memo } from 'react'
import { useFetch } from '../utils/hooks/useFetch'
import { ContextReducer } from './reducer'

export const Context = createContext(null)

export const initialState = {
  target: 'Wizeline',
  theme: false,
  menu: false,
  view: 'home',
  data: [],
}
export const ContextProvider = ({ children }) => {
  const [state, dispatch] = useReducer(ContextReducer, initialState)
  const { data, isLoading } = useFetch(state.target)

  useEffect(() => {
    if (data) dispatch({ type: 'GET_DATA', payload: data })
  }, [data])

  return isLoading ? (
    <div> Loading...</div>
  ) : (
    <Context.Provider value={{ state, dispatch }}> {children} </Context.Provider>
  )
}
export default memo(ContextProvider)
