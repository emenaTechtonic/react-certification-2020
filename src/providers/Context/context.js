import React, { useEffect, createContext, useCallback, useReducer, memo } from 'react'
import { useFetch } from '../../utils/hooks/useFetch'
import { storage } from '../../utils/storage'
import { isElementIn, removeElement } from '../../utils/fns'
import {
  AUTH_STORAGE_USER,
  AUTH_STORAGE_KEY,
  AUTH_STORAGE_FAVORITES,
} from '../../utils/constants'
import { ContextReducer } from './reducer'

export const Context = createContext(null)

export const initialState = {
  target: 'Wizeline',
  theme: false,
  menu: false,
  view: 'home',
  data: [],
  user: { id: 0, name: '', avatarUrl: '' },
  favorites: [],
}
export const ContextProvider = ({ children }) => {
  const [state, dispatch] = useReducer(ContextReducer, initialState)
  const { data } = useFetch(state.target)

  useEffect(() => {
    if (data) dispatch({ type: 'GET_DATA', payload: data })
  }, [data])
  useEffect(() => {
    const lastAuthState = storage.get(AUTH_STORAGE_KEY)
    const isAuthenticated = Boolean(lastAuthState)

    if (isAuthenticated) {
      const lastUser = storage.get(AUTH_STORAGE_USER)
      const userAuthenticated = Object(lastUser)

      dispatch({ type: 'SET_USER', payload: userAuthenticated })

      const lastFavorites = storage.get(AUTH_STORAGE_FAVORITES)
      if (lastFavorites) {
        const favorites = lastFavorites
        dispatch({ type: 'SET_FAVORITES', payload: favorites })
      }
    }
  }, [])

  const saveFavoriteCard = useCallback((card) => {
    let favorites = []
    const lastFavorites = storage.get(AUTH_STORAGE_FAVORITES)
    if (lastFavorites) {
      favorites = lastFavorites
    }
    if (card) {
      const res = isElementIn(card.id, favorites)
      if (!res) {
        favorites.push(card)
        dispatch({ type: 'SET_FAVORITES', payload: favorites })
        storage.set(AUTH_STORAGE_FAVORITES, favorites)
      }
    }
  }, [])
  const removeFavoriteCard = useCallback((card) => {
    let favorites = []
    const lastFavorites = storage.get(AUTH_STORAGE_FAVORITES)
    if (lastFavorites) {
      favorites = lastFavorites
    }
    if (card) {
      const newFavorites = removeElement(card.id, favorites)
      dispatch({ type: 'SET_FAVORITES', payload: newFavorites })
      storage.set(AUTH_STORAGE_FAVORITES, newFavorites)
    }
  }, [])

  // return isLoading ? (
  //   <div> Loading...</div>
  // ) : (
  return (
    <Context.Provider value={{ state, dispatch, saveFavoriteCard, removeFavoriteCard }}>
      {children}
    </Context.Provider>
  )
}
export default memo(ContextProvider)
