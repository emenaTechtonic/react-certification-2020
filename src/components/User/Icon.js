import React, { useContext, useEffect, useState } from 'react'
import { useHistory } from 'react-router-dom'
import { Context } from '../../providers/Context/context'
import { AuthContext } from '../../providers/Auth/auth'
import Portal from '../Portal/Portal'
import Login from '../../pages/Login/Login.page'

const icoLink = 'https://i.pinimg.com/736x/3f/94/70/3f9470b34a8e3f526dbdb022f9f19cf7.jpg'
const user = { id: 0, name: '', avatarUrl: '' }
const Icon = React.memo(() => {
  const history = useHistory()
  const { state, dispatch } = useContext(Context)
  const { authenticated, logout } = useContext(AuthContext)
  const [iconUser, setIconUser] = useState()
  const [isOn, setOn] = useState(false)
  const [coords, setCoords] = useState({})
  const updateLogin = () => {
    setCoords({
      left: window.innerWidth / 2 - 200,
      top: 100,
    })
  }

  useEffect(() => {
    const handleClick = () => {
      if (authenticated) {
        const logoutElement = document.querySelector('.popup')
        logoutElement.classList.remove('login')
        logoutElement.classList.add('logout')
        logoutElement.addEventListener('click', () => {
          logout()
          logoutElement.classList.remove('logout')
          logoutElement.classList.add('login')
          dispatch({ type: 'SET_USER', payload: user })
          history.push('/')
        })
      } else {
        updateLogin()
        setOn(true)
      }
      dispatch({ type: 'OPEN_MENU', payload: false })
    }
    if (!state.user.id) {
      setIconUser(
        <img
          src={icoLink}
          alt={state.user.name}
          width="30"
          height="30"
          className="loginUser"
          onClick={handleClick}
          onKeyDown={handleClick}
          role="presentation"
        />
      )
    } else {
      setIconUser(
        <img
          id="userLoginIcon"
          src={state.user.avatarUrl}
          alt={state.user.name}
          width="30"
          height="30"
          className="loginUser"
          onClick={handleClick}
          onKeyDown={handleClick}
          role="presentation"
        />
      )
      setOn(false)
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [authenticated, state.user.id])
  return (
    <>
      <div className="popup login">
        <ul>
          <li>Logout</li>
        </ul>
      </div>
      {iconUser}
      {isOn && (
        <Portal>
          <Login coords={coords} closeForm={() => setOn(false)} />
        </Portal>
      )}
    </>
  )
})
export default Icon
