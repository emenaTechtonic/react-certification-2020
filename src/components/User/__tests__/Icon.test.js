import React from 'react'
import renderer from 'react-test-renderer'
import { Context } from '../../../providers/Context/context'
import { AuthContext } from '../../../providers/Auth/auth'
import Icon from '../Icon'

const state = {
  target: 'Wizeline',
  theme: false,
  menu: false,
  view: 'home',
  user: { id: 0, name: '', avatarUrl: '' },
}
const login = null
const logout = null
const authenticated = false
describe('Testing Icon component', () => {
  test('Renders Icon without crashing', async () => {
    const rendered = renderer.create(
      <AuthContext.Provider value={{ authenticated, login, logout }}>
        <Context.Provider value={{ state }}>
          <Icon />
        </Context.Provider>
      </AuthContext.Provider>
    )
    expect(rendered).toMatchSnapshot()
  })
})
