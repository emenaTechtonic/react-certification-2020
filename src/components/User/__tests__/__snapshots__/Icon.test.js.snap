// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing Icon component Renders Icon without crashing 1`] = `
<div
  className="popup login"
>
  <ul>
    <li>
      Logout
    </li>
  </ul>
</div>
`;
